README.md
OpenSLMediaPlayer
===============

Re-implementation of Android's [MediaPlayer](http://developer.android.com/reference/android/media/MediaPlayer.html) and [audio effect classes](http://developer.android.com/reference/android/media/audiofx/package-summary.html) based on OpenSL ES APIs.

[![Android Arsenal](https://img.shields.io/badge/Android%20Arsenal-OpenSLMediaPlayer-brightgreen.svg?style=flat)](https://android-arsenal.com/details/1/1434)

Motivation
---

I decided to develop this library to solve these frustrations.

- Android OS does not permit to create extension of [`AudioEffect`](http://developer.android.com/reference/android/media/audiofx/AudioEffect.html) in application layer
- OpenSL ES is a primitive API sets, and it is difficult to use than [`MediaPlayer`](http://developer.android.com/reference/android/media/MediaPlayer.html) class in Java
- [`Visualizer`](http://developer.android.com/reference/android/media/audiofx/Visualizer.html) provides very poor quality captured data (8 bit, monaural, 1024 samples/capture, 20 sampling/second)
- [`Visualizer`](http://developer.android.com/reference/android/media/audiofx/Visualizer.html) does not work properly on some devices
    - [android Issue 64423 "Visualizer with audio session 0 does not always receive all of output mix"](https://code.google.com/p/android/issues/detail?id=64423)
    - [felixpalmer/android-visualizer "Visualization does not work on the new Galaxy devices"](https://github.com/felixpalmer/android-visualizer/issues/5)
    - [Music Visualizer "Limitations on Galaxy S3, S4 and Note 3 (Snapdragon Quad core model)"](https://plus.google.com/110898804228810455198/posts/jXGKDLt9iTz)
    - [Music Visualizer "Limitations on Nexus 4, 5, and 7 (2013)"](https://plus.google.com/110898804228810455198/posts/6chmkb9ix1s)
- Android's [MediaPlayer](http://developer.android.com/reference/android/media/MediaPlayer.html) and [audio effect classes](http://developer.android.com/reference/android/media/audiofx/package-summary.html) are buggy...

Features
---

### Advantages

- Provides both C++ and Java API sets
- Smooth fade in/out when starts/pauses playback
- High quality resampler
- 10 bands graphic equalizer & pre. amplifier
- High quality Visualizer class (floating point, stereo, 32k samples/capture, 60 sampling/second)

### Disadvantages

- **Does not support video playback, audio only** ( Please use [`MediaPlayer`](http://developer.android.com/reference/android/media/MediaPlayer.html) or [`ExoPlayer`](http://developer.android.com/guide/topics/media/exoplayer.html) for video playback)
- Consumes more CPU resources than standard MediaPlayer and other AudioTrack/OpenSL based audio player products (ex. PowerAmp)

### Misc.

- Highly compatibility with standard [`MediaPlayer`](http://developer.android.com/reference/android/media/MediaPlayer.html) and [`audio effect classes`](http://developer.android.com/reference/android/media/audiofx/package-summary.html)
- Implements a lot of workarounds, more better behavior and well-tested ([`Standard*` prefixed API classes](../../wiki/OpenSL-prefixed-API-classes))
- Provides Hybrid media player factory which is a player using OpenSL ES for decoding audio and using `AudioTrack` for playback. This is the most recommended `MediaPlayer` because it provides more tolerance for audio glitches than `OpenSLMediaPlayer` but it can use all features of `OpenSLMediaPlayer`!
- Provides some compatibility class or methods for older devices
- NEON/SSE optimized


Target platforms
---

- Android API level 14+   (since this library depends decoding feature introduced from API level 14)


Latest version
---

- Version 0.7.5  (May 6, 2017)


Demo application (pre-built example apps)
---

See [GitHub Pages site of this project](http://h6ah4i.github.io/android-openslmediaplayer/).


Getting started
---

### Use pre-built library (for Android Studio only)

1. Add these lines to `build.gradle` of your project

```groovy
dependencies {
    compile 'com.h6ah4i.android:openslmediaplayer:0.7.5'
}
```

2. That's all ;)

### Build from source

If you want to build this library from source, please refer to [this article](../../wiki/Build-From-Source).

Documentation
---

Refer to [Wiki](../../wiki) and [JavaDoc](https://h6ah4i.github.io/android-openslmediaplayer/javadoc/0.7.5/).


ToDos
---

- More optimize
- Improve JavaDoc comments


License
---

This library is licensed under the [Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0), and contains some source code files delivered from product of Android Open Source Project (AOSP). See [`AOSP.md`](AOSP.md) for details.

*Note that this library uses some third party libraries, so you also have to take care of their licenses.*

---

### License of OpenSLMediaPlayer (this library)

See [`LICENSE`](LICENSE) for full of the license text.

    Copyright (C) 2014-2015 Haruki Hasegawa

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


---

### License of dependent libraries

#### cxxdasp, cxxporthelper, OpenSL ES CXX

These libraries are licensed under the [Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0).

- [cxxdasp](https://github.com/h6ah4i/cxxdasp)
- [cxxporthelper](https://github.com/h6ah4i/cxxporthelper)
- [lockfreedatastructure](https://github.com/h6ah4i/lockfreedatastructure)
- [OpenSL ES CXX](https://github.com/h6ah4i/openslescxx)


#### PFFFT

This library is licensed under the [FFTPACK5 Software License](https://www2.cisl.ucar.edu/resources/legacy/fft5/license).

- [PFFFT](https://bitbucket.org/jpommier/pffft)
- [PFFFT (h6a h4i forked version)](https://bitbucket.org/h6a_h4i/pffft)


#### Ne10

This library is licensed under the [BSD 3-Clause license](http://opensource.org/licenses/BSD-3-Clause).

- [Ne10](http://projectne10.github.io/Ne10/)


RELEASE-NOTES.md
## 0.7.5

- Fixed monotonic clock not used Android 6.0 or prior devices (issue #35)
- Fixed prepareAsync() broken on StandardMediaPlayer

## 0.7.4

- Strip debug info on AAR (issue #33)
- Fix compilation error with NDK v15

## 0.7.3

- Add workaround for incomplete AudioTrack.write() (issue #26)
- Fix warning message due to inappropriate use of AudioTrack.write() (issue #26)

## 0.7.2

- Fixed 176.4 kHz, 192 kHz audio playback issue
- Fixed build failure issue with the latest NDK (13.0.3315539)

## 0.7.1

- Fixed 11.025 kHz, 12 kHz, 22.05 kHz and 24 kHz audio playback issue (issue #20)
- Update FFTW, libsamplerate, soxr and PFFFT libraries


## 0.7.0

- Added `HybridMediaPlayerFactory` which creates `AudioTrack` audio sink player. It is more tolerant against glitches and strongly recommended than normal `OpenSLMediaPlayer`.
- Fully migrated to Gradle build system. Eclipse is no longer supported.
- Bug fixes


## 0.6.0

- Changed the Standard* prefixed classes not to inherit Android framework classes
- Upgrade Android Gradle build tool version
- Bug fixes
- Code refactoring

## 0.5.0

- Initial release


javadoc/0.7.3/reference/lists.js
var DATA = [
      { id:0, label:"com.h6ah4i.android.media", link:"com/h6ah4i/android/media/package-summary.html", type:"package" },
      { id:1, label:"com.h6ah4i.android.media.IBasicMediaPlayer", link:"com/h6ah4i/android/media/IBasicMediaPlayer.html", type:"class" },
      { id:2, label:"com.h6ah4i.android.media.IBasicMediaPlayer.OnBufferingUpdateListener", link:"com/h6ah4i/android/media/IBasicMediaPlayer.OnBufferingUpdateListener.html", type:"class" },
      { id:3, label:"com.h6ah4i.android.media.IBasicMediaPlayer.OnCompletionListener", link:"com/h6ah4i/android/media/IBasicMediaPlayer.OnCompletionListener.html", type:"class" },
      { id:4, label:"com.h6ah4i.android.media.IBasicMediaPlayer.OnErrorListener", link:"com/h6ah4i/android/media/IBasicMediaPlayer.OnErrorListener.html", type:"class" },
      { id:5, label:"com.h6ah4i.android.media.IBasicMediaPlayer.OnInfoListener", link:"com/h6ah4i/android/media/IBasicMediaPlayer.OnInfoListener.html", type:"class" },
      { id:6, label:"com.h6ah4i.android.media.IBasicMediaPlayer.OnPreparedListener", link:"com/h6ah4i/android/media/IBasicMediaPlayer.OnPreparedListener.html", type:"class" },
      { id:7, label:"com.h6ah4i.android.media.IBasicMediaPlayer.OnSeekCompleteListener", link:"com/h6ah4i/android/media/IBasicMediaPlayer.OnSeekCompleteListener.html", type:"class" },
      { id:8, label:"com.h6ah4i.android.media.IMediaPlayerFactory", link:"com/h6ah4i/android/media/IMediaPlayerFactory.html", type:"class" },
      { id:9, label:"com.h6ah4i.android.media.IReleasable", link:"com/h6ah4i/android/media/IReleasable.html", type:"class" },
      { id:10, label:"com.h6ah4i.android.media.audiofx", link:"com/h6ah4i/android/media/audiofx/package-summary.html", type:"package" },
      { id:11, label:"com.h6ah4i.android.media.audiofx.IAudioEffect", link:"com/h6ah4i/android/media/audiofx/IAudioEffect.html", type:"class" },
      { id:12, label:"com.h6ah4i.android.media.audiofx.IAudioEffect.OnControlStatusChangeListener", link:"com/h6ah4i/android/media/audiofx/IAudioEffect.OnControlStatusChangeListener.html", type:"class" },
      { id:13, label:"com.h6ah4i.android.media.audiofx.IAudioEffect.OnEnableStatusChangeListener", link:"com/h6ah4i/android/media/audiofx/IAudioEffect.OnEnableStatusChangeListener.html", type:"class" },
      { id:14, label:"com.h6ah4i.android.media.audiofx.IBassBoost", link:"com/h6ah4i/android/media/audiofx/IBassBoost.html", type:"class" },
      { id:15, label:"com.h6ah4i.android.media.audiofx.IBassBoost.OnParameterChangeListener", link:"com/h6ah4i/android/media/audiofx/IBassBoost.OnParameterChangeListener.html", type:"class" },
      { id:16, label:"com.h6ah4i.android.media.audiofx.IBassBoost.Settings", link:"com/h6ah4i/android/media/audiofx/IBassBoost.Settings.html", type:"class" },
      { id:17, label:"com.h6ah4i.android.media.audiofx.IEnvironmentalReverb", link:"com/h6ah4i/android/media/audiofx/IEnvironmentalReverb.html", type:"class" },
      { id:18, label:"com.h6ah4i.android.media.audiofx.IEnvironmentalReverb.OnParameterChangeListener", link:"com/h6ah4i/android/media/audiofx/IEnvironmentalReverb.OnParameterChangeListener.html", type:"class" },
      { id:19, label:"com.h6ah4i.android.media.audiofx.IEnvironmentalReverb.Settings", link:"com/h6ah4i/android/media/audiofx/IEnvironmentalReverb.Settings.html", type:"class" },
      { id:20, label:"com.h6ah4i.android.media.audiofx.IEqualizer", link:"com/h6ah4i/android/media/audiofx/IEqualizer.html", type:"class" },
      { id:21, label:"com.h6ah4i.android.media.audiofx.IEqualizer.OnParameterChangeListener", link:"com/h6ah4i/android/media/audiofx/IEqualizer.OnParameterChangeListener.html", type:"class" },
      { id:22, label:"com.h6ah4i.android.media.audiofx.IEqualizer.Settings", link:"com/h6ah4i/android/media/audiofx/IEqualizer.Settings.html", type:"class" },
      { id:23, label:"com.h6ah4i.android.media.audiofx.IHQVisualizer", link:"com/h6ah4i/android/media/audiofx/IHQVisualizer.html", type:"class" },
      { id:24, label:"com.h6ah4i.android.media.audiofx.IHQVisualizer.OnDataCaptureListener", link:"com/h6ah4i/android/media/audiofx/IHQVisualizer.OnDataCaptureListener.html", type:"class" },
      { id:25, label:"com.h6ah4i.android.media.audiofx.ILoudnessEnhancer", link:"com/h6ah4i/android/media/audiofx/ILoudnessEnhancer.html", type:"class" },
      { id:26, label:"com.h6ah4i.android.media.audiofx.ILoudnessEnhancer.Settings", link:"com/h6ah4i/android/media/audiofx/ILoudnessEnhancer.Settings.html", type:"class" },
      { id:27, label:"com.h6ah4i.android.media.audiofx.IPreAmp", link:"com/h6ah4i/android/media/audiofx/IPreAmp.html", type:"class" },
      { id:28, label:"com.h6ah4i.android.media.audiofx.IPreAmp.Settings", link:"com/h6ah4i/android/media/audiofx/IPreAmp.Settings.html", type:"class" },
      { id:29, label:"com.h6ah4i.android.media.audiofx.IPresetReverb", link:"com/h6ah4i/android/media/audiofx/IPresetReverb.html", type:"class" },
      { id:30, label:"com.h6ah4i.android.media.audiofx.IPresetReverb.OnParameterChangeListener", link:"com/h6ah4i/android/media/audiofx/IPresetReverb.OnParameterChangeListener.html", type:"class" },
      { id:31, label:"com.h6ah4i.android.media.audiofx.IPresetReverb.Settings", link:"com/h6ah4i/android/media/audiofx/IPresetReverb.Settings.html", type:"class" },
      { id:32, label:"com.h6ah4i.android.media.audiofx.IVirtualizer", link:"com/h6ah4i/android/media/audiofx/IVirtualizer.html", type:"class" },
      { id:33, label:"com.h6ah4i.android.media.audiofx.IVirtualizer.OnParameterChangeListener", link:"com/h6ah4i/android/media/audiofx/IVirtualizer.OnParameterChangeListener.html", type:"class" },
      { id:34, label:"com.h6ah4i.android.media.audiofx.IVirtualizer.Settings", link:"com/h6ah4i/android/media/audiofx/IVirtualizer.Settings.html", type:"class" },
      { id:35, label:"com.h6ah4i.android.media.audiofx.IVisualizer", link:"com/h6ah4i/android/media/audiofx/IVisualizer.html", type:"class" },
      { id:36, label:"com.h6ah4i.android.media.audiofx.IVisualizer.MeasurementPeakRms", link:"com/h6ah4i/android/media/audiofx/IVisualizer.MeasurementPeakRms.html", type:"class" },
      { id:37, label:"com.h6ah4i.android.media.audiofx.IVisualizer.OnDataCaptureListener", link:"com/h6ah4i/android/media/audiofx/IVisualizer.OnDataCaptureListener.html", type:"class" },
      { id:38, label:"com.h6ah4i.android.media.compat", link:"com/h6ah4i/android/media/compat/package-summary.html", type:"package" },
      { id:39, label:"com.h6ah4i.android.media.compat.AudioAttributes", link:"com/h6ah4i/android/media/compat/AudioAttributes.html", type:"class" },
      { id:40, label:"com.h6ah4i.android.media.compat.AudioAttributes.Builder", link:"com/h6ah4i/android/media/compat/AudioAttributes.Builder.html", type:"class" },
      { id:41, label:"com.h6ah4i.android.media.compat.MediaPlayerCompat", link:"com/h6ah4i/android/media/compat/MediaPlayerCompat.html", type:"class" },
      { id:42, label:"com.h6ah4i.android.media.hybrid", link:"com/h6ah4i/android/media/hybrid/package-summary.html", type:"package" },
      { id:43, label:"com.h6ah4i.android.media.hybrid.HybridMediaPlayerFactory", link:"com/h6ah4i/android/media/hybrid/HybridMediaPlayerFactory.html", type:"class" },
      { id:44, label:"com.h6ah4i.android.media.opensl", link:"com/h6ah4i/android/media/opensl/package-summary.html", type:"package" },
      { id:45, label:"com.h6ah4i.android.media.opensl.OpenSLMediaPlayer", link:"com/h6ah4i/android/media/opensl/OpenSLMediaPlayer.html", type:"class" },
      { id:46, label:"com.h6ah4i.android.media.opensl.OpenSLMediaPlayerContext", link:"com/h6ah4i/android/media/opensl/OpenSLMediaPlayerContext.html", type:"class" },
      { id:47, label:"com.h6ah4i.android.media.opensl.OpenSLMediaPlayerContext.Parameters", link:"com/h6ah4i/android/media/opensl/OpenSLMediaPlayerContext.Parameters.html", type:"class" },
      { id:48, label:"com.h6ah4i.android.media.opensl.OpenSLMediaPlayerFactory", link:"com/h6ah4i/android/media/opensl/OpenSLMediaPlayerFactory.html", type:"class" },
      { id:49, label:"com.h6ah4i.android.media.opensl.OpenSLMediaPlayerNativeLibraryLoader", link:"com/h6ah4i/android/media/opensl/OpenSLMediaPlayerNativeLibraryLoader.html", type:"class" },
      { id:50, label:"com.h6ah4i.android.media.opensl.audiofx", link:"com/h6ah4i/android/media/opensl/audiofx/package-summary.html", type:"package" },
      { id:51, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLBassBoost", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLBassBoost.html", type:"class" },
      { id:52, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLEnvironmentalReverb", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLEnvironmentalReverb.html", type:"class" },
      { id:53, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLEqualizer", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLEqualizer.html", type:"class" },
      { id:54, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLHQEqualizer", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLHQEqualizer.html", type:"class" },
      { id:55, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLHQVisualizer", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLHQVisualizer.html", type:"class" },
      { id:56, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLPreAmp", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLPreAmp.html", type:"class" },
      { id:57, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLPresetReverb", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLPresetReverb.html", type:"class" },
      { id:58, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLVirtualizer", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLVirtualizer.html", type:"class" },
      { id:59, label:"com.h6ah4i.android.media.opensl.audiofx.OpenSLVisualizer", link:"com/h6ah4i/android/media/opensl/audiofx/OpenSLVisualizer.html", type:"class" },
      { id:60, label:"com.h6ah4i.android.media.standard", link:"com/h6ah4i/android/media/standard/package-summary.html", type:"package" },
      { id:61, label:"com.h6ah4i.android.media.standard.StandardMediaPlayer", link:"com/h6ah4i/android/media/standard/StandardMediaPlayer.html", type:"class" },
      { id:62, label:"com.h6ah4i.android.media.standard.StandardMediaPlayerFactory", link:"com/h6ah4i/android/media/standard/StandardMediaPlayerFactory.html", type:"class" },
      { id:63, label:"com.h6ah4i.android.media.standard.audiofx", link:"com/h6ah4i/android/media/standard/audiofx/package-summary.html", type:"package" },
      { id:64, label:"com.h6ah4i.android.media.standard.audiofx.StandardBassBoost", link:"com/h6ah4i/android/media/standard/audiofx/StandardBassBoost.html", type:"class" },
      { id:65, label:"com.h6ah4i.android.media.standard.audiofx.StandardEnvironmentalReverb", link:"com/h6ah4i/android/media/standard/audiofx/StandardEnvironmentalReverb.html", type:"class" },
      { id:66, label:"com.h6ah4i.android.media.standard.audiofx.StandardEqualizer", link:"com/h6ah4i/android/media/standard/audiofx/StandardEqualizer.html", type:"class" },
      { id:67, label:"com.h6ah4i.android.media.standard.audiofx.StandardLoudnessEnhancer", link:"com/h6ah4i/android/media/standard/audiofx/StandardLoudnessEnhancer.html", type:"class" },
      { id:68, label:"com.h6ah4i.android.media.standard.audiofx.StandardPresetReverb", link:"com/h6ah4i/android/media/standard/audiofx/StandardPresetReverb.html", type:"class" },
      { id:69, label:"com.h6ah4i.android.media.standard.audiofx.StandardVirtualizer", link:"com/h6ah4i/android/media/standard/audiofx/StandardVirtualizer.html", type:"class" },
      { id:70, label:"com.h6ah4i.android.media.standard.audiofx.StandardVisualizer", link:"com/h6ah4i/android/media/standard/audiofx/StandardVisualizer.html", type:"class" },
      { id:71, label:"com.h6ah4i.android.media.utils", link:"com/h6ah4i/android/media/utils/package-summary.html", type:"package" },
      { id:72, label:"com.h6ah4i.android.media.utils.AudioEffectSettingsConverter", link:"com/h6ah4i/android/media/utils/AudioEffectSettingsConverter.html", type:"class" },
      { id:73, label:"com.h6ah4i.android.media.utils.AudioSystemUtils", link:"com/h6ah4i/android/media/utils/AudioSystemUtils.html", type:"class" },
      { id:74, label:"com.h6ah4i.android.media.utils.AudioSystemUtils.AudioSystemProperties", link:"com/h6ah4i/android/media/utils/AudioSystemUtils.AudioSystemProperties.html", type:"class" },
      { id:75, label:"com.h6ah4i.android.media.utils.DefaultEqualizerPresets", link:"com/h6ah4i/android/media/utils/DefaultEqualizerPresets.html", type:"class" },
      { id:76, label:"com.h6ah4i.android.media.utils.EnvironmentalReverbPresets", link:"com/h6ah4i/android/media/utils/EnvironmentalReverbPresets.html", type:"class" },
      { id:77, label:"com.h6ah4i.android.media.utils.EqualizerBandInfoCorrector", link:"com/h6ah4i/android/media/utils/EqualizerBandInfoCorrector.html", type:"class" }

    ];
